examples/actor_critic.py

```diff
@@ -0,0 +1,413 @@
+        self.moving_std = MovingExponentialAverage(FLAGS.std_smoothing)
+
+    def forward(self, x):
+        """Forward of both actor and critic"""
+        # Initial layer maps the sequence of one-hot vectors into a
+        # vector of the hidden size. Next layers stay with the same
+        # size and use residual connections.
+        x = F.relu(self.affine1(x))
+        x = x.add(F.relu(self.affine2(x)))
+        x = x.add(F.relu(self.affine3(x)))
+        x = x.add(F.relu(self.affine4(x)))
+
+        # actor: choses action to take from state s_t
+        # by returning probability of each action
+        action_prob = F.softmax(self.action_head(x), dim=-1)
+
+        # critic: evaluates being in the state s_t
+        state_values = self.value_head(x)
+
+        # return values for both actor and critic as a tuple of 2 values:
+        # 1. a list with the probability of each action over the action space
+        # 2. the value from state s_t
+        return action_prob, state_values
+
+
+def select_action(model, state, exploration_rate=0.0):
+    """Selects an action and registers it with the action buffer."""
+    state = torch.from_numpy(state.flatten()).float()
+    probs, state_value = model(state)
+
+    # Create a probability distribution where the probability of
+    # action i is probs[i].
+    m = Categorical(probs)
+
+    # Sample an action using the distribution, or pick an action
+    # uniformly at random if in an exploration mode.
+    if random.random() < exploration_rate:
+        action = torch.tensor(random.randrange(0, len(probs)))
+    else:
+        action = m.sample()
+
+    # Save to action buffer. The drawing of a sample above simply
+    # returns a constant integer that we cannot back-propagate
+    # through, so it is important here that log_prob() is symbolic.
+    model.saved_actions.append(SavedAction(m.log_prob(action), state_value))
+
+    # The action to take.
+    return action.item()
+
+
+def finish_episode(model, optimizer) -> float:
+    """The training code. Calculates actor and critic loss and performs backprop."""
+    R = 0
+    saved_actions = model.saved_actions
+    policy_losses = []  # list to save actor (policy) loss
+    value_losses = []  # list to save critic (value) loss
+    returns = []  # list to save the true values
+
+    # Calculate the true value using rewards returned from the
+    # environment. We are iterating in reverse order while inserting
+    # at each step to the front of the returns list, which implies
+    # that returns[i] is the sum of rewards[j] for j >= i. We do not
+    # use a discount factor as the episode length is fixed and not
+    # very long, but if we had used one, it would appear here.
+    for r in model.rewards[::-1]:
+        R += r
+        returns.insert(0, R)
+
+    # Update the moving averages for mean and standard deviation and
+    # use that to normalize the input.
+    returns = torch.tensor(returns)
+    model.moving_mean.next(returns.mean())
```

Comment:
is the motivation to use moving averages primarily because the episode length is short?

Reply:
The justification for using it is the experiments in the linked pastebin that show that it works better. That may be related to short episode length, and the benefit may fall off for longer episode lengths, I don't know since I didn't try that.

My motivation for trying this in the first place is that data normalization for the input data like this is normally done across the entire dataset at once, so N=everything, which we can't do here since we're generating and training on one data point at a time, so it would be done N=1 as in the official PT example, but N=1 is very different from N=everything and I was thinking that this might make each data point look quite different to the model in a way that it can't sort out. Another way of looking at it is that this is like batchnorm for minibatch=1, where it is known that minibatch should ideally be more like 32 for best results. Doing the normalization even for N=1 is dramatically better than not doing any normalization, though a moving mean works even better.